### PR TITLE
refactor(clerk-js): Pick the redirection prop using a more intuitive approach

### PR DESCRIPTION
**Pull request by desiprisg**
## Type of change

- [ ] 🐛 Bug fix
- [ ] 🌟 New feature
- [ ] 🔨 Breaking change
- [x] 📖 Refactoring / dependency upgrade / documentation
- [ ] other:

## Packages affected

- [ ] `@clerk/clerk-js`
- [ ] `@clerk/clerk-react`
- [ ] `@clerk/nextjs`
- [ ] `@clerk/remix`
- [ ] `@clerk/types`
- [ ] `@clerk/themes`
- [ ] `@clerk/localizations`
- [ ] `@clerk/clerk-expo`
- [ ] `@clerk/backend`
- [ ] `@clerk/clerk-sdk-node`
- [ ] `@clerk/shared`
- [ ] `@clerk/fastify`
- [ ] `@clerk/chrome-extension`
- [ ] `gatsby-plugin-clerk`
- [ ] `build/tooling/chore`

## Description

- [x] `npm test` runs as expected.
- [x] `npm run build` runs as expected.

This is a proposed refactor to make the choosing of the redirect url a more intuitive process. The new function accepts keys and the sources to choose them from, in order. A validator and formatter can also be added.

Initially, this was done as part of a PR that also added some new query params for org actions, but since more logic has been added by the orgs team, I'll leave that up to them to implement. It should be relativelt easy to do after this refactor.

### CODE REVIEW

File: packages/clerk-js/src/ui/contexts/ClerkUIComponentsContext.tsx [hunk 1]
```diff
@@ -41,26 +48,22 @@ export const useSignUpContext = (): SignUpContextType => {
...
-    pickRedirectionProp('afterSignUpUrl', {
-      queryParams,
-      ctx,
-      options,
-      displayConfig,
-    }),
+    pickUrl(['afterSignUpUrl', 'redirectUrl'], queryParams, {
+      validator: url => isAllowedRedirectOrigin(url, options.allowedRedirectOrigins),
+      formatter: camelToSnake,
+    }) || pickUrl(['afterSignUpUrl', 'redirectUrl'], [ctx, options, displayConfig]),
```
[the remark this answers is not in the document]
Comment (desiprisg) [reply packages/clerk-js/src/ui/contexts/ClerkUIComponentsContext.tsx:54]: This is done because we want to add a validator only for the query parameters.

**Comment by SokratisVidros** [comment]
What's the status of this PR? @octoper are you working on it?

File: packages/clerk-js/src/core/clerk.ts [hunk 2]
```diff
@@ -1437,16 +1437,12 @@ export default class Clerk implements ClerkInterface {
     }
 
     const opts: RedirectOptions = {
-      afterSignInUrl: pickRedirectionProp('afterSignInUrl', { ctx: options, options: this.#options }, false),
-      afterSignUpUrl: pickRedirectionProp('afterSignUpUrl', { ctx: options, options: this.#options }, false),
+      afterSignInUrl: pickUrl('afterSignInUrl', [options, this.#options]),
+      afterSignUpUrl: pickUrl('afterSignUpUrl', [options, this.#options]),
```
Comment (dimkl) [inline packages/clerk-js/src/core/clerk.ts:1441]: ❓ @desiprisg is the `pickUrl()` fallbacks to `queryParams.redirect_url` or `options.redirectUrl` that the `pickRedirectionProp` was?

File: packages/clerk-js/src/ui/contexts/ClerkUIComponentsContext.tsx [hunk 3]
```diff
@@ -111,26 +114,22 @@ export const useSignInContext = (): SignInContextType => {
...
-    pickRedirectionProp('afterSignInUrl', {
-      queryParams,
-      ctx,
-      options,
-      displayConfig,
-    }),
+    pickUrl(['afterSignInUrl', 'redirectUrl'], [queryParams], {
+      validator: url => isAllowedRedirectOrigin(url, options.allowedRedirectOrigins),
+      formatter: camelToSnake,
+    }) || pickUrl(['afterSignInUrl', 'redirectUrl'], [ctx, options, displayConfig]),
```
Comment (dimkl) [inline packages/clerk-js/src/ui/contexts/ClerkUIComponentsContext.tsx:127]: ☁️ i think that it's a great idea to refactor the `pickRedirectionProp` to something simpler and more maintainable than currently is but the above code (personal preference) does not make it easier for me to understand the rules of picking the `afterSignInUrl`. :(

**Comment by octoper** [comment]
@SokratisVidros No I accidentaly closed it through another PR.

**Review by nikosdouvlis — CHANGES_REQUESTED** [review]
@desiprisg You will need to apply the changes to the ui-retheme dir.  I will share more details within the day

**Comment by clerk-cookie** [comment]
Hello 👋

We currently close PRs after 60 days of inactivity. It's been 50 days since the last update here. If we missed this PR, please reply here. Otherwise, we'll close this PR in 10 days.

Thanks for being a part of the Clerk community! 🙏

**Comment by clerk-cookie** [comment]
Hello again 👋

After 60 days of no activity, we'll close this PR. Keep in mind, I'm just a robot, so if I've closed this PR in error, please reply here and my human colleagues will reopen it.

Thanks for being a part of the Clerk community! 🙏

**Pull request closed without merging**